stock_available_to_promise_release/models/stock_move.py

```diff
@@ -478,5 +542,35 @@ def _get_all_moves_from_chain(self, field):
                 "move_ids": tuple(self.ids),
             },
         )
-        res.update(dict(self.env.cr.fetchall()))
+        res = defaultdict(dict)
+        all_ids = []
+        for move_id, depth, ids in self.env.cr.fetchall():
+            res[move_id][depth] = ids
+            all_ids += ids
+        if prefetch:
+            self.browse(all_ids)
         return res
+
+    def unrelease(self):
+        """Unrelease moves"""
+        self._check_unrelease_allowed()
+        move_orig_ids = self._get_all_orig_ids(prefetch=True)
+        self.write({"need_release": True})
+        impacted_picking_ids = set()
+        for move in self:
+            move_origs = self.browse(move_orig_ids[move.id]).filtered(
+                lambda m: m.state not in ("done", "cancel")
+            )
+            impacted_picking_ids.update(move_origs.mapped("picking_id").ids)
+            if move_origs:
+                # avoid to propagate cancel to the original move
+                move_origs.write({"propagate_cancel": False})
+                move_origs._action_cancel()
```

Comment:
Needs split before if `move_orig.product_qty` > `move.product_qty`.

Example use case - set a fixed procurement group on pick rule, pick moves are merged
* ship1 <- pack1 <- pickA
* ship2 <- pack2 <- pickA

Unreleasing ship1 needs to split the merged moves of pickA

Reply:
seems to be far more complex.... we could have more than one move as origin and one origin could have more than one move as destination :thinking: